### ISSUE TITLE
The results verified during the training process are inconsistent with the results of the validation script

### ISSUE THREAD

**Issue by Vayne0227** [ultralytics/ultralytics#14929]
### Search before asking

- [X] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

Hello author, I have encountered an issue while training with yolov8-rtdetr. The best mAP50 result for model validation during training was 0.45679, but when I used the validation script for validation, the result was 0.449. Why is the result different.
![image](https://github.com/user-attachments/assets/1517ca3e-d91d-477b-bf5c-3c6106c8ccef)
![image](https://github.com/user-attachments/assets/8d3d52dd-28a8-49d2-a969-07493eb5b0c5)

**Comment by Y-T-G** [comment]
https://github.com/ultralytics/ultralytics/issues/14884#issuecomment-2265129765

**Comment by Vayne0227** [comment]
> [#14884 (comment)](https://github.com/ultralytics/ultralytics/issues/14884#issuecomment-2265129765)

@Y-T-G Hello, I have looked at the above solutions and they did not solve my problem. I did not enable half precision in either method. And during the training process, I also modified the code that was validated, but there were still two situations where the results were consistent.
![image](https://github.com/user-attachments/assets/5073546c-9c7f-4da3-9363-142bd1e4d6b4)

**Comment by Y-T-G** [comment]
The best.pt isn't just based on mAP50 during training.

https://github.com/ultralytics/ultralytics/blob/9f22f45189bd32efc14ddcb7b7b42a548b7aef6f/ultralytics/utils/metrics.py#L743-L746

**Comment by Vayne0227** [comment]
> The best.pt isn't just based on mAP50 during training.
> 
> https://github.com/ultralytics/ultralytics/blob/9f22f45189bd32efc14ddcb7b7b42a548b7aef6f/ultralytics/utils/metrics.py#L743-L746

Thank you for your answer, but I couldn't find the corresponding mAP50 and mAP50:95 results in the CSV file saved during the training process. There is no corresponding 0.449 and 0.259.

**Comment by Y-T-G** [comment]
Run with `half=True`. Find values that correspond to that.

**Comment by Vayne0227** [comment]
> Run with `half=True`. Find values that correspond to that.

I tried your method, but I still haven't found the corresponding result.

**Comment by Y-T-G** [comment]
The values printed are rounded. You can get the correct values from the metrics object.

```python
metrics = model.val()
print(metrics.keys())
```

**Comment by Vayne0227** [comment]
> The values printed are rounded. You can get the correct values from the metrics object.
> 
> ```python
> metrics = model.val(half=True)
> print(metrics.results_dict)
> ```

Thank you very much for your patient answer, but I still haven't found the corresponding result in 'results.csv'. Is this because the calculation method used to save the results during training is different from that used during validation?

**Comment by Y-T-G** [comment]
The only difference would in CSV values are rounded to 5 decimal places, in print, it's rounded to 3 decimal places. So you have to round the results to 5 decimal places and find that.

**Comment by Vayne0227** [comment]
> The only difference would in CSV values are rounded to 5 decimal places, in printed logs, it's rounded to 3 decimal places. So you have to round the results to 5 decimal places and find that.

Yes, I used the rounding method to find the corresponding value earlier, but I couldn't find the corresponding result.

**Comment by pderrenger** [comment]
Thank you for your patience. If rounding isn't revealing the corresponding values, it might be useful to ensure you're using the latest version of the Ultralytics package. Sometimes discrepancies can arise from updates or changes in the codebase. Please update to the latest version and re-run your validation. If the issue persists, consider sharing more details on the specific steps you're taking, so we can further investigate.

**Comment by Vayne0227** [comment]
> Thank you for your patience. If rounding isn't revealing the corresponding values, it might be useful to ensure you're using the latest version of the Ultralytics package. Sometimes discrepancies can arise from updates or changes in the codebase. Please update to the latest version and re-run your validation. If the issue persists, consider sharing more details on the specific steps you're taking, so we can further investigate.

Hello, I am using the latest version of the package. I retrained a network to reproduce the problem I encountered. My training script is as follows:
![image](https://github.com/user-attachments/assets/9b3a7e9a-462b-4319-98ca-b08501255a17)

My validation script is as follows:
![image](https://github.com/user-attachments/assets/2c0dcaf2-eb6f-4721-b665-f2d6c650be1e)

The results retained during the training process are as follows:
![image](https://github.com/user-attachments/assets/0a4bce6c-bd99-49ee-a401-9816ace6ec16)

But the results I obtained when using the validation script for verification are as follows:
![image](https://github.com/user-attachments/assets/d63ce860-857e-4f5d-839b-5a8e161d8235)

I did not use half precision (FP16) during the training and validation process. And during training, I also modified the code for validating the model to not use half precision (FP16). Still unable to find the corresponding result in results.csv.

**Comment by Y-T-G** [comment]
What was the validation result at the end of the training? You shouldn't modify the code.

**Comment by Y-T-G** [comment]
What's the validation result with `batch=1`?

**Comment by Vayne0227** [comment]
> What was the validation result at the end of the training when it run the final evaluation? You shouldn't modify the code.


The results at the end of the training are consistent with those verified using the validation script, but not consistent with the results obtained during the training process. May I ask if I should modify the code used for validation during the training process?
Because I found that during the training process, half precision (FP16) is used to validate the model. Therefore, I made modifications to this part of the [code.](https://github.com/ultralytics/ultralytics/blob/bea4c932783f6ab80cce91e00abe38d45c575531/ultralytics/engine/validator.py#L115)
Ensure that the entire process does not use half precision (FP16)
![image](https://github.com/user-attachments/assets/2c93c0dc-66ce-4c63-90c3-9124b70be311)
I tried batch=1 and the result was 0.436

**Comment by Y-T-G** [comment]
You shouldn't modify it. The model that gets saved is the half precision version. So you will not be able to get the same results as full precision model since that was not saved, even if you set half=False.

**Comment by Vayne0227** [comment]
> You shouldn't modify it. The model that gets saved is the half precision version. So you will not be able to get the same results as full precision model since that was not saved, even if you set half=False. You should use the default training code and run validation with half=True after the training.

Thank you for your answer, but I would like to know how to save the full precision version of the model. Because I experience a loss of NaN when using the half precision (FP16) version of the model.

**Comment by Y-T-G** [comment]
https://github.com/ultralytics/ultralytics/blob/bea4c932783f6ab80cce91e00abe38d45c575531/ultralytics/engine/trainer.py#L492

https://github.com/ultralytics/ultralytics/blob/bea4c932783f6ab80cce91e00abe38d45c575531/ultralytics/engine/model.py#L413

https://github.com/ultralytics/ultralytics/blob/bea4c932783f6ab80cce91e00abe38d45c575531/ultralytics/utils/torch_utils.py#L574

**Comment by Vayne0227** [comment]
> https://github.com/ultralytics/ultralytics/blob/bea4c932783f6ab80cce91e00abe38d45c575531/ultralytics/engine/trainer.py#L492
> 
> https://github.com/ultralytics/ultralytics/blob/bea4c932783f6ab80cce91e00abe38d45c575531/ultralytics/engine/model.py#L413
> 
> https://github.com/ultralytics/ultralytics/blob/bea4c932783f6ab80cce91e00abe38d45c575531/ultralytics/utils/torch_utils.py#L574



I don't understand what you mean. May I ask if I want to delete. half() from these 3 lines of code

**Comment by Y-T-G** [comment]
Yeah

**Comment by Vayne0227** [comment]
> Yeah

Hello, I have resolved your issue. After listening to your feedback, although the final result still does not match the results of the training process, I believe that this approach is very important. Later, I tried to compare the weights of the best model during the training process with the weights of the validated model input, and found that there were some differences in the weights between the two. This resulted in a deviation in the verification results. After passing through the [AutoBacked](https://github.com/ultralytics/ultralytics/blob/5d9046abda3d1070bc2c4bf234a1ab81f50e43e5/ultralytics/engine/validator.py#L124) function, some of the weights of the model have changed. And when I directly use the model without AutoBackend function, the results obtained can correspond to the corresponding results during the training process.

**Comment by pderrenger** [comment]
@Vayne0227 thank you for sharing your findings. It's insightful to know that the differences in weights after passing through the `AutoBackend` function were causing the discrepancies. Using the model without the `AutoBackend` function to match the training results is a valuable approach. If you encounter any further issues or have additional questions, feel free to reach out.
